File: 20_valid_parenthesis.py
class Solution:
    def isValid(self, s: str) -> bool:
        bracket_map = {
            "}" : "{",
            "]" : "[",
            ")" : "("
        }
        brackets = []
        
        for b in s:
            if b in bracket_map:
                if brackets and brackets[-1] == bracket_map[b]:
                    brackets.pop()
                else:
                    return False
            else:
                brackets.append(b)

        return True if not brackets else False
                

File: 20_valid_parenthesis.java
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Map;

class Solution {
    public boolean isValid(String s) {
        Map<Character, Character> bracketMap = Map.of(
             ')', '(',
    '}', '{',
    ']', '['

        );

        Deque<Character> record = new ArrayDeque<>();

        for(char bracket : s.toCharArray()){
            if(bracketMap.containsKey(bracket)){
                if (!record.isEmpty() && bracketMap.get(bracket) == record.peekLast()){
                    record.pollLast();
                } else {
                    return false;
                }
            } else {
                record.addLast(bracket);
            }
        }

        return record.isEmpty();

    }
}